diff --git a/linux/Makefile b/linux/Makefile
index 907a4565..33380938 100644
--- a/linux/Makefile
+++ b/linux/Makefile
@@ -445,6 +445,9 @@ HOST_LFS_CFLAGS := $(shell getconf LFS_CFLAGS 2>/dev/null)
 HOST_LFS_LDFLAGS := $(shell getconf LFS_LDFLAGS 2>/dev/null)
 HOST_LFS_LIBS := $(shell getconf LFS_LIBS 2>/dev/null)
 
+LLVM := $(shell $${CC:-cc} --version | grep -i -q -F clang && echo 1)
+LLVM_IAS := $(LLVM)
+
 ifneq ($(LLVM),)
 ifneq ($(filter %/,$(LLVM)),)
 LLVM_PREFIX := $(LLVM)
@@ -503,7 +506,7 @@ KBUILD_PROCMACROLDFLAGS := $(or $(PROCMACROLDFLAGS),$(KBUILD_HOSTLDFLAGS))
 CPP		= $(CC) -E
 ifneq ($(LLVM),)
 CC		= $(LLVM_PREFIX)clang$(LLVM_SUFFIX)
-LD		= $(LLVM_PREFIX)ld.lld$(LLVM_SUFFIX)
+LD		= $(LLVM_PREFIX)ld$(LLVM_SUFFIX)
 AR		= $(LLVM_PREFIX)llvm-ar$(LLVM_SUFFIX)
 NM		= $(LLVM_PREFIX)llvm-nm$(LLVM_SUFFIX)
 OBJCOPY		= $(LLVM_PREFIX)llvm-objcopy$(LLVM_SUFFIX)
diff --git a/linux/tools/scripts/Makefile.include b/linux/tools/scripts/Makefile.include
index 45f4abef..453096f5 100644
--- a/linux/tools/scripts/Makefile.include
+++ b/linux/tools/scripts/Makefile.include
@@ -60,7 +60,7 @@ endif
 
 $(call allow-override,CC,$(LLVM_PREFIX)clang$(LLVM_SUFFIX))
 $(call allow-override,AR,$(LLVM_PREFIX)llvm-ar$(LLVM_SUFFIX))
-$(call allow-override,LD,$(LLVM_PREFIX)ld.lld$(LLVM_SUFFIX))
+$(call allow-override,LD,$(LLVM_PREFIX)ld$(LLVM_SUFFIX))
 $(call allow-override,CXX,$(LLVM_PREFIX)clang++$(LLVM_SUFFIX))
 $(call allow-override,STRIP,$(LLVM_PREFIX)llvm-strip$(LLVM_SUFFIX))
 else
diff --git a/u-boot/Makefile b/u-boot/Makefile
index 3a5e0452..ec0c7fd9 100644
--- a/u-boot/Makefile
+++ b/u-boot/Makefile
@@ -392,13 +392,8 @@ include scripts/Kbuild.include
 # Make variables (CC, etc...)
 
 AS		= $(CROSS_COMPILE)as
-# Always use GNU ld
-ifneq ($(shell $(CROSS_COMPILE)ld.bfd -v 2> /dev/null),)
-LD		= $(CROSS_COMPILE)ld.bfd
-else
 LD		= $(CROSS_COMPILE)ld
-endif
-CC		= $(CROSS_COMPILE)gcc
+CC		= $(CROSS_COMPILE)cc
 CPP		= $(CC) -E
 AR		= $(CROSS_COMPILE)ar
 NM		= $(CROSS_COMPILE)nm
