diff --git a/linux/Makefile b/linux/Makefile
index 61f8019e..bccb3444 100644
--- a/linux/Makefile
+++ b/linux/Makefile
@@ -443,6 +443,9 @@ HOST_LFS_CFLAGS := $(shell getconf LFS_CFLAGS 2>/dev/null)
 HOST_LFS_LDFLAGS := $(shell getconf LFS_LDFLAGS 2>/dev/null)
 HOST_LFS_LIBS := $(shell getconf LFS_LIBS 2>/dev/null)
 
+LLVM := $(shell $${CC:-cc} --version | grep -i -q -F clang && echo 1)
+LLVM_IAS := $(LLVM)
+
 ifneq ($(LLVM),)
 ifneq ($(filter %/,$(LLVM)),)
 LLVM_PREFIX := $(LLVM)
@@ -514,7 +517,7 @@ KBUILD_PROCMACROLDFLAGS := $(or $(PROCMACROLDFLAGS),$(KBUILD_HOSTLDFLAGS))
 CPP		= $(CC) -E
 ifneq ($(LLVM),)
 CC		= $(LLVM_PREFIX)clang$(LLVM_SUFFIX)
-LD		= $(LLVM_PREFIX)ld.lld$(LLVM_SUFFIX)
+LD		= $(LLVM_PREFIX)ld$(LLVM_SUFFIX)
 AR		= $(LLVM_PREFIX)llvm-ar$(LLVM_SUFFIX)
 NM		= $(LLVM_PREFIX)llvm-nm$(LLVM_SUFFIX)
 OBJCOPY		= $(LLVM_PREFIX)llvm-objcopy$(LLVM_SUFFIX)
diff --git a/linux/tools/scripts/Makefile.include b/linux/tools/scripts/Makefile.include
index b5ecf137..ebc0b360 100644
--- a/linux/tools/scripts/Makefile.include
+++ b/linux/tools/scripts/Makefile.include
@@ -60,7 +60,7 @@ endif
 
 $(call allow-override,CC,$(LLVM_PREFIX)clang$(LLVM_SUFFIX))
 $(call allow-override,AR,$(LLVM_PREFIX)llvm-ar$(LLVM_SUFFIX))
-$(call allow-override,LD,$(LLVM_PREFIX)ld.lld$(LLVM_SUFFIX))
+$(call allow-override,LD,$(LLVM_PREFIX)ld$(LLVM_SUFFIX))
 $(call allow-override,CXX,$(LLVM_PREFIX)clang++$(LLVM_SUFFIX))
 $(call allow-override,STRIP,$(LLVM_PREFIX)llvm-strip$(LLVM_SUFFIX))
 else
diff --git a/u-boot/arch/arm/cpu/armv8/u-boot.lds b/u-boot/arch/arm/cpu/armv8/u-boot.lds
index 231ab28b..ff1659ee 100644
--- a/u-boot/arch/arm/cpu/armv8/u-boot.lds
+++ b/u-boot/arch/arm/cpu/armv8/u-boot.lds
@@ -154,7 +154,7 @@ SECTIONS
 	/*
 	 * arch/arm/lib/crt0_64.S assumes __bss_start - __bss_end % 8 == 0
 	 */
-	.bss ADDR(.rela.dyn) (OVERLAY) : {
+	.bss ALIGN(8) : {
 		__bss_start = .;
 		*(.bss*)
 		. = ALIGN(8);
diff --git a/u-boot/drivers/clk/mediatek/clk-mtk.c b/u-boot/drivers/clk/mediatek/clk-mtk.c
index b4de3871..485c5205 100644
--- a/u-boot/drivers/clk/mediatek/clk-mtk.c
+++ b/u-boot/drivers/clk/mediatek/clk-mtk.c
@@ -36,7 +36,7 @@
 
 /* shared functions */
 
-static const int mtk_common_clk_of_xlate(struct clk *clk,
+static int mtk_common_clk_of_xlate(struct clk *clk,
 					 struct ofnode_phandle_args *args,
 					 const struct mtk_clk_tree *tree)
 {
@@ -316,7 +316,7 @@ static void mtk_clk_print_mux_parents(struct mtk_clk_priv *priv,
 
 /* apmixedsys functions */
 
-static const int mtk_apmixedsys_of_xlate(struct clk *clk,
+static int mtk_apmixedsys_of_xlate(struct clk *clk,
 					 struct ofnode_phandle_args *args)
 {
 	struct mtk_clk_priv *priv = dev_get_priv(clk->dev);
@@ -590,7 +590,7 @@ static void mtk_apmixedsys_dump(struct udevice *dev)
 
 /* topckgen functions */
 
-static const int mtk_topckgen_of_xlate(struct clk *clk,
+static int mtk_topckgen_of_xlate(struct clk *clk,
 				       struct ofnode_phandle_args *args)
 {
 	struct mtk_clk_priv *priv = dev_get_priv(clk->dev);
@@ -903,7 +903,7 @@ static void mtk_topckgen_dump(struct udevice *dev)
 
 /* infrasys functions */
 
-static const int mtk_infrasys_of_xlate(struct clk *clk,
+static int mtk_infrasys_of_xlate(struct clk *clk,
 				       struct ofnode_phandle_args *args)
 {
 	struct mtk_clk_priv *priv = dev_get_priv(clk->dev);
@@ -1083,7 +1083,7 @@ static void mtk_infrasys_dump(struct udevice *dev)
 
 /* CG functions */
 
-static const int mtk_clk_gate_of_xlate(struct clk *clk,
+static int mtk_clk_gate_of_xlate(struct clk *clk,
 				       struct ofnode_phandle_args *args)
 {
 	struct mtk_cg_priv *priv = dev_get_priv(clk->dev);
